Mathlib/AlgebraicGeometry/EllipticCurve/NormalForms.lean

```diff
@@ -0,0 +1,661 @@
+  rw [self.c₆]
+  linear_combination -288 * W.a₆ * CharP.cast_eq_zero R 3
+
+theorem Δ_of_char_three [CharP R 3] : W.Δ = -W.a₄ ^ 3 := by
+  rw [self.Δ]
+  linear_combination (-21 * W.a₄ ^ 3 - 144 * W.a₆ ^ 2) * CharP.cast_eq_zero R 3
+
+end IsShortNF
+
+namespace IsShortNF
+
+variable {E} (self : E.IsShortNF)
+include self
+
+theorem j : E.j = 6912 * E.a₄ ^ 3 / (4 * E.a₄ ^ 3 + 27 * E.a₆ ^ 2) := by
+  have h := E.Δ'.ne_zero
+  rw [E.coe_Δ', self.Δ] at h
+  rw [EllipticCurve.j, Units.val_inv_eq_inv_val, ← div_eq_inv_mul, E.coe_Δ',
+    self.c₄, self.Δ, div_eq_div_iff h (right_ne_zero_of_mul h)]
+  ring1
+
+theorem j_of_char_three [CharP F 3] : E.j = 0 := by
+  simp [EllipticCurve.j, self.c₄_of_char_three]
+
+end IsShortNF
+
+section VariableChange
+
+variable [Invertible (2 : R)] [Invertible (3 : R)]
+
+/-- There is an explicit change of variables of a `WeierstrassCurve` to
+a short normal form, provided that 2 and 3 are invertible in the ring.
+It is the composition of an explicit change of variables with `WeierstrassCurve.toCharNeTwoNF`. -/
+def toShortNF : VariableChange R :=
+  ⟨1, ⅟3 * -(W.variableChange W.toCharNeTwoNF).a₂, 0, 0⟩ * W.toCharNeTwoNF
+
+theorem toShortNF_spec : (W.variableChange W.toShortNF).IsShortNF := by
+  erw [variableChange_comp]
+  constructor <;> simp
+
+theorem exists_variableChange_isShortNF :
+    ∃ C : VariableChange R, (W.variableChange C).IsShortNF :=
+  ⟨_, W.toShortNF_spec⟩
+
+end VariableChange
+
+/-! ### Normal forms of characteristic = 3 and j ≠ 0 -/
+
+/-- A `WeierstrassCurve` is in normal form of characteristic = 3 and j ≠ 0, if its
+$a_1, a_3, a_4 = 0$. In other words it is $Y^2 = X^3 + a_2X^2 + a_6$. -/
+@[mk_iff]
+structure IsCharThreeJNeZeroNF : Prop where
+  a₁ : W.a₁ = 0
+  a₃ : W.a₃ = 0
+  a₄ : W.a₄ = 0
+
+namespace IsCharThreeJNeZeroNF
+
+variable {W} (self : W.IsCharThreeJNeZeroNF)
+include self
+
+theorem isCharNeTwoNF : W.IsCharNeTwoNF := ⟨self.a₁, self.a₃⟩
+
+theorem b₂ : W.b₂ = 4 * W.a₂ := self.isCharNeTwoNF.b₂
+
+theorem b₄ : W.b₄ = 0 := by
+  simp [self.a₄, self.isCharNeTwoNF.b₄]
+
+theorem b₆ : W.b₆ = 4 * W.a₆ := self.isCharNeTwoNF.b₆
+
+theorem b₈ : W.b₈ = 4 * W.a₂ * W.a₆ := by
+  simp [self.a₄, self.isCharNeTwoNF.b₈]
+
+theorem c₄ : W.c₄ = 16 * W.a₂ ^ 2 := by
+  simp [self.a₄, self.isCharNeTwoNF.c₄]
+
+theorem c₆ : W.c₆ = -64 * W.a₂ ^ 3 - 864 * W.a₆ := by
+  simp [self.a₄, self.isCharNeTwoNF.c₆]
+
+theorem Δ : W.Δ = -64 * W.a₂ ^ 3 * W.a₆ - 432 * W.a₆ ^ 2 := by
+  simp [self.a₄, self.isCharNeTwoNF.Δ]
+
+theorem c₄_of_char_three [CharP R 3] : W.c₄ = W.a₂ ^ 2 := by
+  rw [self.c₄]
+  linear_combination 5 * W.a₂ ^ 2 * CharP.cast_eq_zero R 3
+
+theorem c₆_of_char_three [CharP R 3] : W.c₆ = -W.a₂ ^ 3 := by
+  rw [self.c₆]
+  linear_combination (-21 * W.a₂ ^ 3 - 288 * W.a₆) * CharP.cast_eq_zero R 3
+
+theorem Δ_of_char_three [CharP R 3] : W.Δ = -W.a₂ ^ 3 * W.a₆ := by
+  rw [self.Δ]
+  linear_combination (-21 * W.a₂ ^ 3 * W.a₆ - 144 * W.a₆ ^ 2) * CharP.cast_eq_zero R 3
```

Comment:
I guess you can add b2, b6, and b8 too here for completeness? Same for the others for `_char_two`.

Reply:
Added.